Weighted trajectories

pmenczel opened the pull request:
**Description**

Introduces weights for the trajectories of trajectory solvers. These weights enter all trajectory averages. This concept is a generalization of two things that are already in qutip:
* The influence martingale in `nm_mcsolve` is simply a time-dependent weight from this point of view
* The "improved sampling" option for `mcsolve` weights the no-jump trajectory with its exact probability, and then only samples from the ensemble of trajectories with at least one jump.

Currently, these two things cannot be combined: "improved sampling" does not work for `nm_mcsolve`. With these changes, it would work.
In the future, there might be at least one more application for weighted trajectories:
* For mixed initial states in `mcsolve`, it might be useful to allow sampling of initial states with frequencies or probabilities that are different from their true prefactor in the initial state. The ratio between the sampling frequencies and true prefactors would be weights on the trajectories.

This also fixes some things in merging `McResult`s and `NmmcResult`s, and perhaps fixes the target tolerance computation for the "improved sampling" option.

**Implementation**

A difficulty in the implementation is that the weights depend on the total number of trajectories in different ways. For example, the no-jump trajectory has a fixed weight that never changes, but all other trajectories come with prefactors $(1-p_0) / (N-1)$ where $p_0$ is the probability of the no-jump trajectory, and $N$ the total number of trajectories (including the one no-jump trajectory). For this reason, I separate weights into *absolute weights* (like $p_0$) and *relative weights* (like the others). The average is performed as follows:

$$ \bar\rho = \sum_{T: \text{abs}} w_a(T) w_r(T) \rho(T) + \frac{1}{N_{\text{rel}}} \sum _{T: \text{rel}} w_r(T) \rho(T) $$

where the first sum is over all trajectories T that have absolute weights, and the second sum over all other trajectories. Here, $w_a(T)$ and $w_r(T)$ are the absolute and relative weights, $\rho(T)$ the state associated with a trajectory, and $N_{\text{rel}}$ the number of trajectories that have not been assigned absolute weights.

Intuitively, the idea is that absolute weights are for trajectories where the contribution to the full state is known exactly, and the trajectories without absolute weights are the stochastic part.

**Things to do**
- [x] Update tests and create new ones
- [x] Update photocurrent to work with weights
- [x] Update target tolerance calculation to work with weights
- [x] Think about how result objects should be added if both contain trajectories with absolute weights
- [x] Double-check that this creates no problems for stochastic solvers
- [x] Towncrier entry

**Related issues or PRs**
https://github.com/qutip/qutip/issues/2235

Ericgig commented on qutip/solver/result.py:
```diff
@@ -1046,6 +1201,8 @@ def photocurrent(self):
         """
         Average photocurrent or measurement of the evolution.
         """
+        # TODO this is wrong if trajectories have weights
+        # unclear how to implement in case of time-dependent weights
```
Does photocurrent make sense for nm_mcsolve?
If it does not, we could raise an error when used with time-dependent weights.

pmenczel replied:
It probably does not make sense for nm_mcsolve. I have added an error in the case of time-dependent weights, and removed the tests for it.

Ericgig commented on qutip/solver/result.py:
```diff
@@ -894,78 +919,208 @@ def __repr__(self):
 
         new = self.__class__(
-            self._raw_ops, self.options, solver=self.solver, stats=self.stats
+            self.e_ops, self.options, solver=self.solver, stats=self.stats
```
`e_ops` are callable wrapped in `ExpectOp`, by passing them here, they get wrapped twice and raw_ops are no longer available.

Ericgig commented on qutip/solver/result.py:
```diff
@@ -894,78 +919,208 @@ def __repr__(self):
     def __add__(self, other):
         if not isinstance(other, MultiTrajResult):
-            return NotImplemented
+            raise NotImplementedError("Can only add two multi traj results")
```
Why not let python raise the error when no appropriate functions is found?
Not that I think this would happen, it stop other project to develop something compatible with our results.

pmenczel replied:
Sorry, this was a mistake, reverted the change

Ericgig reviewed (commented):
Thank you for merging the results classes.

The default seems to be a relative weight of `1`. Meaning that it should work fine in solver which does not use it (stochastic). But should they just keep using `Result` until we have a use for those weight?

It there a balancing tools for weight values?
If there are 2 runs, can both have a relative weight of 10 or 0.1 without affecting the end results? If there are 2 starting kets that we want to merge with a weight of 10:1, does the actual values matter or only the ratio?

I feel it would be safer to compute the relative weight as `W(T) / sum_T(W(T))` instead of using the number of trajectories.

I am not sure how it works with weight changing over time...

pmenczel commented:
Thank you for your review.

> The default seems to be a relative weight of `1`. Meaning that it should work fine in solver which does not use it (stochastic). But should they just keep using `Result` until we have a use for those weight?

I let them use `Result` again for now. I needed to add some lines [here](https://github.com/pmenczel/qutip/blob/weighted-trajectories/qutip/solver/result.py#L710) so that the new `MultiTrajResult` can deal with it. (Which might be a good idea anyway.)

> It there a balancing tools for weight values? If there are 2 runs, can both have a relative weight of 10 or 0.1 without affecting the end results? If there are 2 starting kets that we want to merge with a weight of 10:1, does the actual values matter or only the ratio?

I have added a `merge` function that lets the user merge trajectory results with custom relative weights. I think trajectories with absolute or relative weights are handled consistently. Calling `result1 + result2` uses default values for the relative weights, chosen such that every trajectory contributes equally.

> I feel it would be safer to compute the relative weight as `W(T) / sum_T(W(T))` instead of using the number of trajectories.

Hmm unfortunately I think that this won't work. For the "improved sampling" case, the sum of the relative weights is not one.

Ericgig commented on qutip/solver/result.py:
```diff
@@ -895,7 +970,31 @@ def __repr__(self):
         return "\n".join(lines)
 
-    def __add__(self, other):
+    def merge(self, other, p=None):
```
The `__add__` is already in the users hand and part of the official interface.
It is documented in the [guide](https://qutip.readthedocs.io/en/qutip-5.0.x/guide/dynamics/dynamics-monte.html#changing-the-number-of-trajectories) (and maybe the tutorials).
It's new, so it's probably not used much, but still we shouldn't remove it without a deprecation warning or a major release...

It would be better to keep it as merging with the default value of `p`.

Ericgig reviewed (commented):
It's nice to merge `McResultImprovedSampling` into `McResult`, but the overall complexity has increased... I believe it would be better to move `MultiTrajResult` to it's own file.

Did you review the `dynamics-nmmonte,rst`? Adding a note/example of using the improved sampling option there would be good.

pmenczel commented:
> It's nice to merge `McResultImprovedSampling` into `McResult`, but the overall complexity has increased... I believe it would be better to move `MultiTrajResult` to it's own file.
> 
> Did you review the `dynamics-nmmonte,rst`? Adding a note/example of using the improved sampling option there would be good.

Thank you for the review. The complexity has increased, more than I expected when I started writing. I hope it is worth it?
I have extracted the `MultiTrajResult` and its subclasses to a new module, `qutip.solver.multitrajresult`.

I have added a mention about the improved sampling option in the guide, and I will have a look at the tutorial notebook whether it makes sense to include somewhere there.

pmenczel commented on qutip/solver/stochastic.py:
```diff
@@ -165,6 +165,12 @@ def wiener_process(self):
         """
         return self._trajectories_attr("wiener_process")
 
+    def merge(self, other, p=None):
+        raise NotImplementedError("Merging results of the stochastic solvers "
+                                  "is currently not supported. Please raise "
+                                  "an issue on GitHub if you would like to "
+                                  "see this feature.")
+
```
Currently, merging the `measurement`, `dW` etc attributes of stochastic results doesn't work. That was the reason why I removed merge here, but maybe that went a bit too far.

Should I just remove this, or maybe leave in a warning in the case a user tries to merge stochastic results where `keep_measurements` was used?

pmenczel commented on qutip/tests/solver/test_results.py:
```diff
@@ -352,3 +459,115 @@ def test_merge_result(self, keep_runs_results):
+                random_state = qutip.rand_ket(dim, seed=seed)
+                traj.add(t, random_state)
+
+            if time_dep_weights and np.random.randint(2):
```
I wanted to add weights to only some of the trajectories, to test that nothing goes wrong if some of the traj don't have weights.

Ericgig reviewed (approved):
I think the increased complexity is fine in this case.
Everything looks good. Thank you.
